### PR TITLE
feat: Add Bagaaravel record permissions serializer mixin

### PR DESCRIPTION
**Pull request by bertdeblock**

### CODE REVIEW

File: tests/dummy/config/optional-features.json [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-  "jquery-integration": false
+  "jquery-integration": true
```
Comment (Windvis) [inline tests/dummy/config/optional-features.json]: 🤮

Comment (bertdeblock) [reply tests/dummy/config/optional-features.json]: 😂 needed for ED during tests.

File: addon/mixins/serializer.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    if (Object.keys(payloadPermissions).length === 0) {
+      return normalized
+    }
+
+    let modelName = modelClass.modelName
+    let recordId = payload.id
+    let recordPermissions = {}
+
+    modelClass.fields.forEach((kind, fieldName) => {
```
Comment (Windvis) [inline addon/mixins/serializer.js]: I would split this in a separate method, this has nothing to do with normalizing anymore.

File: addon/mixins/serializer.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      let fieldKey
+
+      if (kind === 'attribute') {
+        fieldKey = this.keyForAttribute(fieldName)
+      } else {
+        fieldKey = this.keyForRelationship(fieldName)
+      }
+
+      if (typeof payloadPermissions[fieldKey] !== 'string') {
```
Comment (Windvis) [inline addon/mixins/serializer.js]: I would store this in a well named variable to make it more clear why you are doing this.

File: addon/mixins/serializer.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  }
+
+  return {}
+}
+
+function getBitwisePermissions (stringPermissions) {
+  let bitwisePermissions = PERMISSIONS.NONE
+
+  if (stringPermissions.charAt(0) === 'r') {
```
Comment (Windvis) [inline addon/mixins/serializer.js]: Does the order of the permission actually matter? `rw` vs `wr`?

Comment (bertdeblock) [reply addon/mixins/serializer.js]: It does for OS read/write/execute (rwx) permissions for example, so I thought I'd enforce it here as well. I want the API to be consistent.

File: addon/mixins/serializer.js (hunk 2 above)
Comment (bertdeblock) [reply addon/mixins/serializer.js]: What exactly do you mean by `has nothing to do with normalizing anymore`?

File: addon/mixins/serializer.js (hunk 3 above)
Comment (bertdeblock) [reply addon/mixins/serializer.js]: Will do

File: addon/mixins/serializer.js (hunk 2 above)
Comment (Windvis) [reply addon/mixins/serializer.js]: `The normalize method is used to convert a payload received from your external data source into the normalized form store.push() expects. You should override this method, munge the hash and return the normalized payload.` [Source](https://api.emberjs.com/ember-data/3.8/classes/DS.Serializer/methods/normalize?anchor=normalize)

This kind of logic has nothing to do with that, right? We just use the method as a hook for our own logic. To make things more readable I would split that up into a separate method which can be called from the `normalize` method. I might be wrong though.

File: addon/mixins/serializer.js (hunk 4 above)
Comment (Windvis) [reply addon/mixins/serializer.js]: Fair enough

File: tests/dummy/config/optional-features.json (hunk 1 above)
Comment (Windvis) [reply tests/dummy/config/optional-features.json]: Can't we set it somewhere else? Specifically for the tests?

Comment (Windvis) [reply tests/dummy/config/optional-features.json]: https://github.com/Bagaar/ember-data-record-permissions/blob/7c2ff045130a07fc3e276d06415a6ef625d0a15b/config/ember-try.js#L17 
Looks like we can do something similar by adding it to environment.js file, but don't bother. It's probably also needed outside of the tests?

File: addon/mixins/serializer.js (hunk 2 above)
Comment (Windvis) [reply addon/mixins/serializer.js]: `modelClass.fields.forEach` also seems to fail on 2.18

File: addon/mixins/serializer.js (hunk 3 above)
Comment (bertdeblock) [reply addon/mixins/serializer.js]: Updated.

File: addon/mixins/serializer.js (hunk 2 above)
Comment (bertdeblock) [reply addon/mixins/serializer.js]: Updated, still need to check the `modelClass.fields.forEach` though.

**Comment by bertdeblock** [comment]
Closing in favour of !9

**Pull request closed without merging**